### DISCUSSION TITLE
PreToolkit — Composable, reproducible data preprocessing

### DISCUSSION THREAD

**Discussion by akifkulll** [0ctacity/.github#18] [Ideas]
I'm proposing an open-source Python library for the .LUCID team called **PreToolkit**.

The idea is to make data preprocessing composable, reproducible, and maintainable — a set of building blocks you snap together rather than code you rewrite in every project.

**The problem**

In almost every ML project, a large share of the work is preparing the data: filling missing values, encoding categoricals, scaling features. Yet this same code gets rewritten from scratch in project after project, in slightly different ways each time. Worse, it's easy to fall into traps — the most common being data leakage, where information from the test set silently influences the training process because a statistic was fit on the wrong data.

Existing tools don't fully close this gap. scikit-learn provides the primitives, but leaves reproducibility, data-type awareness, and feature-name tracking up to the user. Some libraries that tried to sit on top of it (e.g. `pretools`) took a different direction — automated feature selection — and have been unmaintained since 2020.

**The approach**

PreToolkit would treat each preprocessing step as a self-contained, chainable block. The core would care about three things scikit-learn does not make central:

- **Reproducibility first** — every block records its fitted state and the exact transformation applied, so the same dataset yields identical output across runs and people.
- **Data-type awareness** — blocks know which columns are numeric vs. categoric and route transformations automatically, instead of the user slicing column indices by hand.
- **Metadata tracking** — the toolkit tracks how input feature names map to output feature names across the whole chain, which is painful to do manually after one-hot encoding or binning.

**The architecture**

A single abstract base class defines the contract; every block inherits from it and shares the same interface.

```
BasePreprocessor (abstract)
  ├── fit()  ·  transform()  ·  fit_transform()   ← shared logic
  └── _fit() ·  _transform()                      ← each block implements

Concrete blocks:
  ├── MissingValueImputer   (mean / median / most_frequent)
  ├── CategoricalEncoder    (one-hot / label / ordinal)
  └── FeatureScaler         (standard / minmax / robust)

Pipeline (also a BasePreprocessor):
  └── chains blocks, applied in sequence
```

Because all blocks share `fit` / `transform`, the pipeline treats them uniformly (polymorphism), and new blocks can be added without touching existing code. The public `fit()` handles shared concerns — including a leakage guard, so `fit` runs on training data only while `transform` applies to both train and test.

Usage would look like:

```
pipe = Pipeline([
    MissingValueImputer(strategy="median"),
    CategoricalEncoder(method="onehot"),
    FeatureScaler(method="standard"),
])

X_train_clean = pipe.fit_transform(X_train)
X_test_clean  = pipe.transform(X_test)   # no fit — no leakage
```

**Scope**

The first release (v0.1) would stay deliberately small: the three blocks above, a pipeline to chain them, leakage protection, unit tests, and a clear README with an end-to-end example on a messy dataset. Everything beyond that — ordinal encoding, robust scaling, outlier handling, binning, advanced encoders for high-cardinality columns — becomes an open contribution opportunity, so the project grows over time rather than trying to do everything at once.

**Why this fits**

It matches .LUCID's focus on reproducible AI systems, it's a developer tool rather than a one-off demo, and its modular design makes it easy for several people to contribute in parallel — each owning a different block without blocking anyone else.

**Implementation**

Python, with numpy and pandas as the core dependencies (scikit-learn optional, wrapped internally where useful). Standard packaging via `pyproject.toml`, tests with pytest, CI on every PR, MIT licensed — consistent with other Octacity repos.

I'd love to hear thoughts — especially on the base-class interface, how much we lean on scikit-learn internally vs. staying independent, the feature-name mapping approach, and where the line should sit between v0.1 and later contributions.

All contributions are welcome.
